pkg/controller/stack/stack_controller.go

```diff
@@ -1085,6 +1086,85 @@ func (sess *reconcileStackSession) SetEnvRefsForWorkspace(ctx context.Context, w
 	return nil
 }
 
+func (sess *reconcileStackSession) resolveConfigRefs(ctx context.Context) ([]ConfigKeyValue, error) {
+	allConfigs := make([]ConfigKeyValue, 0)
+	for _, configRefMap := range sess.stack.ConfigRefs {
+		for k, ref := range configRefMap {
+			// ConfigMap and Structured are special config cases, so they are checked first
+			switch ref.SelectorType {
+			case shared.ConfigResourceSelectorConfigMap:
+				configMapRef := ref.ConfigMapRef
+				if configMapRef != nil {
+					var config corev1.ConfigMap
+					if err := sess.kubeClient.Get(ctx, types.NamespacedName{Name: configMapRef.Name, Namespace: configMapRef.Namespace}, &config); err != nil {
+						return nil, fmt.Errorf("Failed to get the ConfigMap %s on namespace %s: %w", configMapRef.Name, configMapRef.Namespace, err)
+					}
+					// assumes the whole configmaps's data is the config content; try to read as a conventional stack yaml config
+					var configMapContent map[string]any
+					if err := yaml.Unmarshal([]byte(config.Data[configMapRef.Key]), &configMapContent); err != nil {
+						return nil, fmt.Errorf("Failed to read the ConfigMap content as a stack YAML config. Namespace=%s Name=%s: %w", configMapRef.Namespace, configMapRef.Name, err)
+					}
+					structuredConfig := StructuredConfig(configMapContent).Flatten()
+					allConfigs = append(allConfigs, structuredConfig...)
+				}
+			case shared.ConfigResourceSelectorLiteral:
+				literalRef := ref.ConfigLiteralRef
+				if literalRef != nil {
+					// ConfigLiteralRef handles both simple and structured values as json, flattening all keys to build a list of Pulumi key:value configs
+					structuredConfig, err := NewStructuredConfigFromJSON(k, literalRef.Value)
+					if err != nil {
+						return nil, fmt.Errorf("Failed to unmarshall %s as a structured config: %w", k, err)
+					}
+					configs := structuredConfig.Flatten()
+					allConfigs = append(allConfigs, configs...)
+				}
+			// Secret should be handled here as well because auto.ConfigValue should be marked as Secret:true
+			case shared.ConfigResourceSelectorType(shared.ResourceSelectorSecret):
+				secretValue, err := sess.resolveResourceRef(ctx, &shared.ResourceRef{
+					SelectorType: shared.ResourceSelectorSecret,
+					ResourceSelector: shared.ResourceSelector{
+						FileSystem: ref.FileSystem,
+						Env:        ref.Env,
+						SecretRef:  ref.SecretRef,
+					},
+				})
+				if err != nil {
+					return nil, err
+				}
+				allConfigs = append(allConfigs, ConfigKeyValue{
+					Key: k,
+					Value: auto.ConfigValue{
+						Value:  secretValue,
+						Secret: true,
+					},
+				})
+			default:
+				// try to resolve as a ResourceRef
```

Comment:
Do we need this? `resolveResourceRef` doesn't support structure so it seems odd to default to it. An error would be reasonable IMO.

Reply:
It doesn't, but `resolveResourceRef` supports all other cases like `env`, `file`, etc. These options are supported in `configRefs` too. 

Maybe it does not make sense and `configRefs` should support just structured/config map options?

Reply:
Yes, we should keep the new API surface area as small as possible. Unless we have a good reason to include them now, we can omit `env`, `file` etc. and add them later if needed. We don't need a key `k` for the literal case which leaves this default case as the only reason for it. Not needed IMO.